lib/remote_storage/s3.rb

```diff
@@ -0,0 +1,110 @@
+require "remote_storage/rest_provider"
+require "digest"
+require "base64"
+require "openssl"
+require "webrick/httputils"
+
+module RemoteStorage
+  class S3
+    include RestProvider
+
+    private
+
+    # S3 already wraps the ETag around quotes
+    def format_etag(etag)
+      etag
+    end
+
+    def do_put_request(url, data, content_type)
+      deal_with_unauthorized_requests do
+        md5 = Digest::MD5.base64digest(data)
+        authorization_headers = authorization_headers_for("PUT", url, md5, content_type)
+        RestClient.put(url, data, authorization_headers.merge({ "Content-Type" => content_type, "Content-Md5" => md5}))
```

Comment:
This line is too long. Let's break the object and put all keys and values on seperate lines for readability.